sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1025,7 +1025,8 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
     // state and all the input rows processed before. In another word, the order of input rows
     // matters for non-deterministic expressions, while pushing down predicates changes the order.
     case filter @ Filter(condition, project @ Project(fields, grandChild))
-      if fields.forall(_.deterministic) =>
+        if fields.forall(_.deterministic) &&
+          fields.forall(_.find(_.isInstanceOf[ScalaUDF]).isEmpty) =>
```

Comment:
I'm not sure if I understand this correctly, do you mean `ScalaUDF` can be nondeterministic and we should always treat it as nondeterministic expression? If so, I think a better idea is just override `deterministic` in `ScalaUDF` and always return false.